File: leetcode/86-partition-list.cpp
#include <cstddef>
#include <iostream>

using namespace std;

struct ListNode {
  int val;
  ListNode *next;

  ListNode() : val(0), next(NULL) {}
  ListNode(int x) : val(x), next(NULL) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

void printList(ListNode *head) {
  ListNode *temp = head;
  while (temp) {
    cout << temp->val;

    if (temp->next) {
      cout << " | ";
    }

    temp = temp->next;
  }
  cout << endl;
}

class Solution {

public:
  ListNode *head;
  Solution() { this->head = NULL; }

  ListNode *append(int value) {
    ListNode *newNode = new ListNode(value);

    if (this->head == NULL) {
      head = newNode;
    } else {
      ListNode *currentNode = head;
      while (currentNode->next != NULL) {
        currentNode = currentNode->next;
      }
      currentNode->next = newNode;
    }

    return newNode;
  }

  ListNode *partition(ListNode *head, int x) {
    if (head == NULL) {
      return head;
    }
    if (head->next == NULL) {
      return head;
    }

    ListNode *temp = head;

    ListNode *smallPrev = new ListNode();
    ListNode *smallHead = smallPrev;

    ListNode *bigPrev = new ListNode();
    ListNode *bigHead = bigPrev;

    while (temp) {
      // such a chame, i had to clone the node... how can i achieve without a
      // copy? is it possible
      ListNode *current = new ListNode(temp->val) if (temp->val < x) {
        smallPrev->next = current;
        smallPrev = current;
      }
      else {
        bigPrev->next = current;
        bigPrev = current;
      }

      ListNode *toDelete = temp; // memory clean up
      temp = temp->next;
      delete toDelete; // memory clean up
    }

    smallHead = smallHead->next;
    bigHead = bigHead->next;

    smallPrev->next = bigHead;

    if (smallHead == NULL) {
      return bigHead;
    }
    return smallHead;
  }
};

int main() {
  Solution solution;

  // ListNode *head = solution.append(3);
  // solution.append(8);
  // solution.append(5);
  // solution.append(10);
  // solution.append(2);
  // solution.append(1);

  // printList(head);
  // ListNode *result = solution.partition(solution.head, 5);
  // printList(result);

  ListNode *head = solution.append(1);
  solution.append(1);

  printList(head);
  ListNode *result = solution.partition(solution.head, 0);
  printList(result);

  cout << "------------------------------" << endl;
  cout << "Are you test reading this correctly?" << endl;
}

File: leetcode/86-partition-list.js
class ListNode {
  val;
  next;

  constructor(val = 0, next = null) {
    this.val = val;
    this.next = next;
  }
}

function printList(head) {
  temp = head;
  while (temp) {
    console.log(temp.val);

    if (temp.next) {
      console.log(" | ");
    }

    temp = temp.next;
  }
  console.log("\n");
}

class Solution {
  append(value) {
    const newNode = new ListNode(value);

    if (this.head == null) {
      this.head = newNode;
    } else {
      let currentNode = head;
      while (currentNode.next != null) {
        currentNode = currentNode.next;
      }
      currentNode.next = newNode;
    }

    return newNode;
  }

  // the edge cases are killing me.
  // input: [1,1] x: 0 output: [1,1]
  // input: [1,1] x: 2 output: [1,1]
  partition(head, x) {
    if (head == null) {
      return head;
    }

    if (head.next == null) {
      // if only one item always return it
      return head;
    }

    let temp = head;

    let smallHead = temp.val < x ? temp : null;
    let smallTail = temp.val < x ? temp : null;

    let bigHead = temp.val >= x ? temp : null;

    let tempNext = temp.next;

    while (temp.next) {
      tempNext = temp.next.next;
      if (temp.next.val < x) {
        if (smallHead == null) {
          smallHead = temp.next;
          smallTail = smallHead;
        }

        smallTail.next = temp.next;
        smallTail = temp.next;

        temp.next = tempNext;
        smallTail.next = null;

        continue;
      } else {
        if (bigHead == null) {
          bigHead = temp.next;
        }
      }

      temp = temp.next;
    }

    console.log("--------------------");
    const smallHeadValue = smallHead != null ? smallHead.val : -1;
    console.log("small head: ", smallHeadValue);

    const smallTailValue = smallTail != null ? smallTail.val : -1;
    console.log("small tail: ", smallTailValue);

    const bigHeadValue = bigHead != null ? bigHead.val : -1;
    console.log("big head: ", bigHeadValue);

    const headValue = head != null ? head.val : -1;
    console.log("head: ", headValue);

    const smallHeadNextValue =
      smallHead != null && smallHead.next != null ? smallHead.next.val : -1;
    console.log("small head next: ", smallHeadNextValue);

    console.log("--------------------");

    if (smallTail == null) {
      // when there's no small tail it means all items were bigger than x and we
      // do not have a small list
      return bigHead;
    }

    // merge both partitions, the small at the start and the big at the end!
    smallTail.next = bigHead;

    // then we return the smallHead which is the newest head that contains
    return smallHead;
  }
}

const solution = new Solution();

// head = solution.append(3);
// solution.append(8);
// solution.append(5);
// solution.append(10);
// solution.append(2);
// solution.append(1);

// printList(head);
// result = solution.partition(solution.head, 5);
// printList(result);

// head = solution.append(0);
// solution.append(1);

// printList(head);
// result = solution.partition(solution.head, 2);
// printList(result);

console.log("------------------------------");
console.log("Are you reading this correctly?");
